### PR TITLE
Miscellaneous fixes to the x-transformers implementation

### PR DESCRIPTION
**Pull request by Waino**
- Validation no longer crashes (transposes were missing).
- A distributed component covering the parameters in the `TransformerWrapper` object, most notably `to_logits`.
- Arguments of `TransformerWrapper` can be set through the config file.
- A fix to the content of state dicts, avoiding duplicate storage of some parameters.
- Removal of some obsolete opts.
- Correctly handle stats both with and without accuracy computation (type of initial value is inferred from preceding stats object).

### CODE REVIEW

File: mammoth/model_builder.py [hunk 1]
```diff
@@ -310,3 +330,21 @@ def build_model(
     # logger.info(model)
     logger.info('Building model - done!')
     return model
+
+
+def validate_optimizer_coverage(model, optimizer):
```
Comment (TimotheeMickus) [inline mammoth/model_builder.py:335]: docstring

File: mammoth/distributed/components.py [hunk 2]
```diff
@@ -87,7 +87,50 @@ def needs_communication(self) -> bool:
         return self.group is not None
 
 
-# TODO: This is a misnomer: Not an entire XCoder, but just one AttentionLayers block
+@dataclass  # type: ignore
+class DistributedTransformerWrapper(DistributedComponent, ABC):
```
Comment (TimotheeMickus) [inline mammoth/distributed/components.py:91]: docstring. It's fairly self-evident but making it clear what is the purpose of this cls is is bound to help the next person to work on is

File: mammoth/distributed/components.py [hunk 3]
```diff
@@ -87,7 +87,50 @@ def needs_communication(self) -> bool:
...
+    task_id: str
+    side: Side
+
+    def get_name(self) -> str:
+        return f'{self.side.name}_{self.task_id}'
+
+    def get_module(self, model: NMTModel) -> nn.Module:
+        parent = model.encoder if self.side == Side.encoder else model.decoder
+        tw = parent[self.task_id]
```
Comment (TimotheeMickus) [inline mammoth/distributed/components.py:100]: tw is a pretty unclear variable name

File: mammoth/distributed/components.py [hunk 4]
```diff
@@ -87,7 +87,50 @@ def needs_communication(self) -> bool:
...
+        mismatch = module.load_state_dict(state_dict, strict=False)
+        missing_keys = [
+            name for name in mismatch.missing_keys
+            if not name.startswith('attn_layers.') or name.startswith('token_emb.')
+        ]
+        return mismatch._replace(missing_keys=missing_keys)
+
+
 @dataclass  # type: ignore
 class DistributedAttentionLayersBlock(DistributedComponent, ABC):
```
Comment (TimotheeMickus) [inline mammoth/distributed/components.py:135]: docstring

File: mammoth/distributed/components.py [hunk 5]
```diff
@@ -106,22 +149,32 @@ def named_parameters(self, model: NMTModel):
...
+        missing_keys = [
+            name for name in mismatch.missing_keys
+            if not name.startswith('layers.')
+        ]
+        return mismatch._replace(missing_keys=missing_keys)
+
 
 @dataclass
-class DistributedEncoder(DistributedAttentionLayersBlock):
+class DistributedEncoderAttentionLayersBlock(DistributedAttentionLayersBlock):
```
Comment (TimotheeMickus) [inline mammoth/distributed/components.py:177]: docstring

File: mammoth/distributed/components.py [hunk 6]
```diff
@@ -136,7 +189,7 @@ def get_module(self, model: NMTModel) -> nn.Module:
 
 
 @dataclass
-class DistributedDecoder(DistributedAttentionLayersBlock):
+class DistributedDecoderAttentionLayersBlock(DistributedAttentionLayersBlock):
```
Comment (TimotheeMickus) [inline mammoth/distributed/components.py:192]: docstring

File: mammoth/modules/transformer_decoder.py [hunk 7]
```diff
@@ -317,9 +303,9 @@ def from_opts(cls, opts, embeddings, is_on_top=False):
             embeddings,
             opts.max_relative_positions,
             opts.aan_useffn,
-            opts.full_context_alignment,
-            opts.alignment_layer,
-            alignment_heads=opts.alignment_heads,
+            False,
```
Comment (TimotheeMickus) [inline mammoth/modules/transformer_decoder.py:306]: random hardcoded values in the middle of a list of args is not cool...

3 options:
- kwargs would help. 
- add a comment if you're forcing default values, maybe add a log if the default value conflicts with the user reqs? 
- remove the corresponding args in the constructor

File: mammoth/opts.py [hunk 8]
```diff
@@ -255,7 +255,11 @@ def model_opts(parser):
         "For more detailed information, see: "
         "https://arxiv.org/pdf/1803.02155.pdf",
     )
-    group.add('--heads', '-heads', type=int, default=8, help='Number of heads for transformer self-attention')
+    group.add(
+        '--heads', '-heads', type=int, default=8,
+        help='Number of heads for transformer self-attention. '
+        ' Semi-obsolete: not used for x-transformers, only used for some attention bridge configuations.'
```
Comment (TimotheeMickus) [inline mammoth/opts.py:261]: maybe worth renaming to ab_heads or stg

File: mammoth/opts.py [hunk 9]
```diff
@@ -395,6 +370,10 @@ def _add_train_general_opts(parser):
         default=None,
         help='Criteria to use for early stopping.',
     )
+    group.add(
+        '--max_nan_batches', '-max_nan_batches', type=int, default=5,
```
Comment (TimotheeMickus) [inline mammoth/opts.py:374]: what's the logic for the 5 default?

File: mammoth/opts.py [hunk 10]
```diff
@@ -554,7 +533,7 @@ def _add_train_general_opts(parser):
         '--max_grad_norm',
         '-max_grad_norm',
         type=float,
-        default=5,
+        default=1,
```
Comment (TimotheeMickus) [inline mammoth/opts.py:536]: default float if type float?

File: mammoth/opts.py [hunk 11]
```diff
@@ -677,7 +656,7 @@ def _add_train_general_opts(parser):
         '-learning_rate',
         type=float,
         default=1.0,
-        help="Starting learning rate. Recommended settings: sgd = 1, adagrad = 0.1, adadelta = 1, adam = 0.001",
+        help="Starting learning rate. Recommended settings: sgd = TBD, adagrad = TBD, adadelta = TBD, adam = TBD",
```
Comment (TimotheeMickus) [inline mammoth/opts.py:659]: comment out the recommendation for now?

File: mammoth/trainer.py [hunk 12]
```diff
@@ -381,28 +386,40 @@ def validate(self, valid_iter, moving_average=None, task=None):
...
+                # else:
+                #     normalization = batch.batch_size
 
                 with torch.cuda.amp.autocast(enabled=self.optim.amp):
                     # F-prop through the model.
                     logits, decoder_output = valid_model(
-                        src,
-                        decoder_input,
-                        src_lengths,
+                        rearrange(src, 't b 1 -> b t'),
```
Comment (TimotheeMickus) [inline mammoth/trainer.py:404]: in a perfect world we would just normalize the tensors to batch seq dim across the lib, but this world isn't perfect.

File: mammoth/trainer.py [hunk 13]
```diff
@@ -476,6 +493,9 @@ def _gradient_accumulation(
 
             try:
                 if loss is not None:
+                    if torch.isnan(loss):
+                        raise Exception('Loss blowout')
```
Comment (TimotheeMickus) [inline mammoth/trainer.py:497]: can we/should we type this?

File: mammoth/trainer.py [hunk 14]
```diff
@@ -496,10 +516,13 @@ def _gradient_accumulation(
                 total_stats.update(batch_stats)
                 report_stats.update(batch_stats)
                 report_stats.update_task_loss(batch_stats.loss, metadata)
-
             except Exception:
```
Comment (TimotheeMickus) [inline mammoth/trainer.py:519]: i think this is also meant to to catch cuda oom's, isn't it? in which case your nan counter isn't valid?

**Review by TimotheeMickus — APPROVED** [review]
i mostly have nitpickery of the docstring variety.

File: mammoth/modules/transformer_decoder.py (hunk 7 above)
Comment (Waino) [reply mammoth/modules/transformer_decoder.py:306]: I actually want to remove this. The old transformer implementation is no longer needed for anything other than the attention bridge.

I think I can safely remove the decoder, as the AB only needs TransformerEncoderLayer and MultiHeadedAttention. I'm not completely sure what depends on what, so I didn't yet remove anything.

I'm not going to port the attention bridge to x-transformers. As far as I'm concerned, it is obsolete.

Comment (TimotheeMickus) [reply mammoth/modules/transformer_decoder.py:306]: I don't think we have any reasons (outside of the sentimental) to keep the AB in the impl; but that feels like a discussion with Jörg and Raúl.

File: mammoth/opts.py (hunk 9 above)
Comment (TimotheeMickus) [reply mammoth/opts.py:374]: honestly, I'd rather have the default zero and give the flag a scary name — some degree of flexibility is useful, but this is essentially an invitation to brute force a buggy impl / poor hparam choices

File: mammoth/trainer.py (hunk 12 above)
Comment (Waino) [reply mammoth/trainer.py:404]: Indeed. #80

File: mammoth/trainer.py (hunk 13 above)
Comment (Waino) [reply mammoth/trainer.py:497]: I think typed exceptions are mainly useful for recoverable problems that you may want to catch somewhere higher up. Blowing out of the training loop is not going to be recoverable.

File: mammoth/trainer.py (hunk 14 above)
Comment (Waino) [reply mammoth/trainer.py:519]: Yeah, that's right. The current code maybe allows recovering from a single slightly too large batch, by catching the OOM. This is quite unlikely, though, and wasn't intentional.

We could have a typed exception that increases the nan counter, and let everything else through directly. That would cause OOMs to be instantly fatal.

File: mammoth/trainer.py (hunk 13 above)
Comment (TimotheeMickus) [reply mammoth/trainer.py:497]: I express a strong disagreement.

1. You do catch it (and everything else) on line 519. 
2. If you want a clean break, then make it a break / branching out cond. If you want a weird stack jump behavior, make it clear to the other devs you're doing a custom hack around a potential problem 
3. I would also argue typing helps  other devs: e.g., assume my dev crashes because of a typo producing a out of bounds, not being caught by this try/except here is bound to make my stack less of pain

The minimal thing is to just make it a runtime so as to avoid cases as in 3, but i wouldn't be against finding something more exotic / custom, that way you could except ExoticError to deal with nan blowouts and except RuntimeError for cuda ooms

Comment (Waino) [reply mammoth/trainer.py:497]: You are absolutely right.

I already forgot what I was trying to do here. The catch-everything-and-retry-forever was already there from before, and I tried to make it less bad, when I should have just removed it.

I'll remove the try/except, and the `--max_nan_batches` opt.
